### PR TITLE
Add Direct Method Check to Leaf Device.

### PR DESCRIPTION
**Pull request by aribeironovaes**

### CODE REVIEW

File: smoke/LeafDevice/details/Details.cs [hunk 1]
```diff
@@ -185,6 +216,7 @@ protected Task MaybeDeleteDeviceIdentity()
     public class DeviceContext
     {
         public Device Device;
+        public DeviceClient DeviceClientInstance;
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: Why is this needed?

File: smoke/LeafDevice/details/Details.cs [hunk 2]
```diff
@@ -104,6 +106,7 @@ async Task CreateDeviceIdentity(RegistryManager rm)
             this.context = new DeviceContext
             {
                 Device = device,
+                DeviceClientInstance = null,
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: This doesn't seem like a good idea..

File: smoke/LeafDevice/details/Details.cs [hunk 3]
```diff
@@ -156,6 +159,34 @@ protected async Task VerifyDataOnIoTHub()
...
+            Console.WriteLine("Leaf device received direct method call...");
+            return Task.FromResult(new MethodResponse((int)HttpStatusCode.OK));
+        }
+
+        protected async Task VerifyDirectMethod()
+        {
+            //User Service SDK to invoke Direct Method on the device.
+            ServiceClient serviceClient =
+                ServiceClient.CreateFromConnectionString(this.context.IotHubConnectionString);
+
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: Nit: Remove extra white lines..

File: smoke/LeafDevice/details/Details.cs [hunk 4]
```diff
@@ -156,6 +159,34 @@ protected async Task VerifyDataOnIoTHub()
...
+
+            //Call a direct method
+            using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(300)))
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: I wonder if using the ConnectionTimeout and ResponseTimeout is a better check instead of the Cancellation token here..

File: smoke/LeafDevice/details/Details.cs [hunk 5]
```diff
@@ -156,6 +159,34 @@ protected async Task VerifyDataOnIoTHub()
             await eventHubClient.CloseAsync();
         }
 
+        static Task<MethodResponse> DirectMethod(MethodRequest methodRequest, object userContext)
+        {
+            Console.WriteLine("Leaf device received direct method call...");
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: While we are testing it, maybe we should pass a payload and return a payload here. That will be a better check imo..

File: smoke/LeafDevice/LeafDevice.cs [hunk 6]
```diff
@@ -27,6 +27,7 @@ public async Task RunAsync()
                 await GetOrCreateDeviceIdentity();
                 await ConnectToEdgeAndSendData();
                 await this.VerifyDataOnIoTHub();
+                await this.VerifyDirectMethod();
```
Comment (varunpuranik) [inline smoke/LeafDevice/LeafDevice.cs]: Are we planning to add more checks here, like GetTwin, update reported properties, etc?

Comment (aribeironovaes) [reply smoke/LeafDevice/LeafDevice.cs]: Not right now. The possibilities are endless. For this PBI this is the last scenario, but I encourage Mike to keep an ongoing E2E Effort so we can add more coverage.

File: smoke/LeafDevice/details/Details.cs [hunk 7]
```diff
@@ -185,6 +216,7 @@ protected Task MaybeDeleteDeviceIdentity()
     public class DeviceContext
```
[the remark this answers is not in the document]
Comment (aribeironovaes) [reply smoke/LeafDevice/details/Details.cs]: Understand, but want to keep this change minimal, without refactoring. 

---
In reply to: [219919372](https://github.com/Azure/iotedge/pull/344#discussion_r219919372) [](ancestors = 219919372)

File: smoke/LeafDevice/details/Details.cs (hunk 3 above)
Comment (aribeironovaes) [reply smoke/LeafDevice/details/Details.cs]: Done. 

---
In reply to: [219917970](https://github.com/Azure/iotedge/pull/344#discussion_r219917970) [](ancestors = 219917970)

File: smoke/LeafDevice/details/Details.cs (hunk 2 above)
Comment (varunpuranik) [reply smoke/LeafDevice/details/Details.cs]: Option<DeviceClient>

File: smoke/LeafDevice/details/Details.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (varunpuranik) [reply smoke/LeafDevice/details/Details.cs]: I don't think the change will be that big.. but either ways, I can send a cleanup PR if you want..

File: smoke/LeafDevice/details/Details.cs (hunk 5 above)
Comment (aribeironovaes) [reply smoke/LeafDevice/details/Details.cs]: Good idea. Done. Testing. 

---
In reply to: [219918895](https://github.com/Azure/iotedge/pull/344#discussion_r219918895) [](ancestors = 219918895)

File: smoke/LeafDevice/details/Details.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aribeironovaes) [reply smoke/LeafDevice/details/Details.cs]: Done. 

---
In reply to: [219926565](https://github.com/Azure/iotedge/pull/344#discussion_r219926565) [](ancestors = 219926565)

File: smoke/LeafDevice/details/Details.cs (hunk 2 above)
Comment (aribeironovaes) [reply smoke/LeafDevice/details/Details.cs]: Thanks! 

---
In reply to: [219926180](https://github.com/Azure/iotedge/pull/344#discussion_r219926180) [](ancestors = 219926180)
